db/tags.go

```diff
@@ -96,17 +98,98 @@ func RemoveTags(imageID int64, tags []common.Tag) error {
 	})
 }
 
-// Attempt to complete a tag by suggesting up to 10 possible tags for a prefix
+// Try to match an incomplete tag against a list of postfixes, return matches
+func matchPost(re *regexp.Regexp, prefix string,
+	postfixes []string) (matches []string) {
+	for _, p := range postfixes {
+		matched := re.FindString(p)
+		if len(matched) != 0 {
+			matches = append(matches, prefix+matched)
+		}
+	}
+	return
+}
+
+// Attempt to complete a tag by suggesting up to 20 possible tags for a prefix
 func CompleteTag(s string) (tags []string, err error) {
 	const maxCap = 20
 	tags = make([]string, 0, maxCap)
+	typeQ := ""
+	prefix := ""
+	if s[0] == '-' {
+		prefix = "-"
+		s = s[1:]
+	}
 
-	// TODO: Inject system tags into array
+	i := strings.IndexByte(s, ':')
+	if i != -1 {
+		// Complete postfix
+		re := regexp.MustCompile(`^` + regexp.QuoteMeta(s[i+1:]) + `.*`)
+		switch s[:i] {
+		case "undefined":
+			prefix += s[:i+1]
+			s = s[i+1:]
+			typeQ = fmt.Sprintf(" and type = %d", common.Undefined)
```

Comment:
Please move all cases of this into an embedded function to DRY this a bit. 
Something like 
```
completeWithPrefix := func (t common.TagType) {
  prefix += s[:i+1]
  s = s[i+1:]
  typeQ = fmt.Sprintf(" and type = %d", t)
}
```